### PR TITLE
Speed up isbitsunion and bitsunionsize functions

### PR DESCRIPTION
**Pull request by quinnj**
Comparison of implementations for `Base.isbitsunion` and `Base.bitsunionsize`:
```julia
julia> function isbitsunion(U::Union)
           for u in Base.uniontypes(U)
               isbits(u) || return false
           end
           return true
       end
isbitsunion (generic function with 2 methods)

julia> isbitsunion(T) = false
isbitsunion (generic function with 2 methods)

julia> isbitsunion2(u::Union) = isbitsunion2(u.a) & isbitsunion2(u.b)
isbitsunion2 (generic function with 1 method)

julia> isbitsunion2(x) = isbits(x)
isbitsunion2 (generic function with 2 methods)

julia> isbitsunion3(u) = ccall(:jl_islayout_inline, Bool, (Any, Ref{Csize_t}, Ref{Csize_t}), u, Ref{Csize_t}(), Ref{Csize_t}())
isbitsunion3 (generic function with 1 method)

julia> b1 = @benchmarkable isbitsunion(Union{Int8, UInt8, Int16, UInt16, Int32, UInt32, Int64, UInt64, String})
julia> tune!(b1)
julia> run(b1)
BenchmarkTools.Trial:
  memory estimate:  352 bytes
  allocs estimate:  4
  --------------
  minimum time:     1.311 μs (0.00% GC)
  median time:      1.369 μs (0.00% GC)
  mean time:        1.655 μs (7.85% GC)
  maximum time:     1.303 ms (99.70% GC)
  --------------
  samples:          10000
  evals/sample:     10

julia> b12 = @benchmarkable isbitsunion2(Union{Int8, UInt8, Int16, UInt16, Int32, UInt32, Int64, UInt64, String})
Benchmark(evals=1, seconds=5.0, samples=10000)

julia> tune!(b12)
Benchmark(evals=10, seconds=5.0, samples=10000)

julia> run(b12)
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     1.258 μs (0.00% GC)
  median time:      1.274 μs (0.00% GC)
  mean time:        1.452 μs (0.00% GC)
  maximum time:     9.890 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     10

julia> b13 = @benchmarkable isbitsunion3(Union{Int8, UInt8, Int16, UInt16, Int32, UInt32, Int64, UInt64, String})
Benchmark(evals=1, seconds=5.0, samples=10000)

julia> tune!(b13)

Benchmark(evals=987, seconds=5.0, samples=10000)

julia> run(b13)
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     48.988 ns (0.00% GC)
  median time:      49.077 ns (0.00% GC)
  mean time:        53.574 ns (0.00% GC)
  maximum time:     312.398 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     987

julia> function bitsunionsize(U::Union)
           sz = 0
           for u in Base.uniontypes(U)
               sz = max(sz, sizeof(u))
           end
           return sz
       end
bitsunionsize (generic function with 1 method)

julia> function bitsunionsize2(U::Union)
           sz = Ref{Csize_t}()
           algn = Ref{Csize_t}()
           @assert ccall(:jl_islayout_inline, Bool, (Any, Ref{Csize_t}, Ref{Csize_t}), U, sz, algn)
           return sz[]
       end
bitsunionsize2 (generic function with 1 method)

julia> b1 = @benchmarkable bitsunionsize(Union{Int8, UInt8, Int16, UInt16, Int32, UInt32, Int64, UInt64})
Benchmark(evals=1, seconds=5.0, samples=10000)

julia> tune!(b1)
Benchmark(evals=10, seconds=5.0, samples=10000)

julia> run(b1)
BenchmarkTools.Trial:
  memory estimate:  208 bytes
  allocs estimate:  3
  --------------
  minimum time:     1.437 μs (0.00% GC)
  median time:      1.467 μs (0.00% GC)
  mean time:        1.563 μs (0.00% GC)
  maximum time:     8.141 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     10

julia> b12 = @benchmarkable bitsunionsize2(Union{Int8, UInt8, Int16, UInt16, Int32, UInt32, Int64, UInt64})
Benchmark(evals=1, seconds=5.0, samples=10000)

julia> tune!(b12)
Benchmark(evals=989, seconds=5.0, samples=10000)

julia> run(b12)
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     45.847 ns (0.00% GC)
  median time:      45.939 ns (0.00% GC)
  mean time:        49.987 ns (0.00% GC)
  maximum time:     274.160 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     989
```

### CODE REVIEW

**Comment by quinnj** [comment]
@vtjnash, any concerns w/ this approach?

File: base/array.jl [hunk 1]
```diff
@@ -114,27 +114,19 @@ asize_from(a::Array, n) = n > ndims(a) ? () : (arraysize(a,n), asize_from(a, n+1
...
-function isbitsunion end
-
-function isbitsunion(U::Union)
-    for u in Base.uniontypes(U)
-        isbits(u) || return false
-    end
-    return true
-end
-isbitsunion(T) = false
+isbitsunion(u::Union) = ccall(:jl_islayout_inline, Bool, (Any, Ref{Csize_t}, Ref{Csize_t}), u, Ref{Csize_t}(), Ref{Csize_t}())
```
Comment (vtjnash) [inline base/array.jl]: more typically called as `jl_array_store_unboxed`, I think

Comment (vtjnash) [reply base/array.jl]: return value is a `Cint`, not `Bool`

**Comment by vtjnash** [comment]
This is OK, but doesn't implement the same function anymore. In particular, while the `jl_islayout_inline` query is related to `isbitsunion`, it is not correct to assume the queries are equivalent.

**Pull request merged**